File: EDUCATIVE/String_Problems/word_pattern.py
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        >>> pattern="abba"
        >>> str="dog cat cat dog"
        >>> str = str.split()
        >>> str
        ['dog', 'cat', 'cat', 'dog']
        >>> a = zip(pattern, str)
        >>> a
        [('a', 'dog'), ('b', 'cat'), ('b', 'cat'), ('a', 'dog')]
        >>> set(a)
        set([('b', 'cat'), ('a', 'dog')])
        >>> set(str)
        set(['dog', 'cat'])
        >>> set(pattern)
        set(['a', 'b'])
        """
        str = str.split()
        a = zip(pattern, str)
        print a
        return len(pattern) == len(str) and 
        len(set(a)) == len(set(pattern)) == len(set(str))

File: EDUCATIVE/String_Problems/word_pattern.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean wordPattern(String pattern, String str) {

        String [] words = str.split(" ");

        if(words.length != str.length()) 
            return false;
        
        Map<Character, String> index = new HashMap<>();

        for(int i=0; i < words.length; i++) {
           if(index.containsKey(pattern.charAt(i))){
               //System.out.print(index.get(pattern.charAt(i)));
                if (!index.get(pattern.charAt(i)).equals(words[i])) {
                    return false;
                }
           } else {
               if(map.containsValue(words[i])) {
                   return false;
               }
               index.put(pattern.charAt(i), words[i]);
               System.out.print(index.get(pattern.charAt(i)));
           }
           index.put(pattern.charAt(i), words[i]);
           //System.out.print(index.get(pattern.charAt(i)));
            
           
        }
        return true;
    }
}



